Convert the Log widget to use custom View/Model

jagannatharjun opened the pull request:
following #12144

jagannatharjun commented:
`QPainter::drawText` seems to work but just to be sure I'm now using `QStyle::drawItemText` in `LogItemDelegate::paint`

![image](https://user-images.githubusercontent.com/34717789/78880911-b2ee2280-7a73-11ea-98c7-aba69cbe3fe5.png)

glassez commented on src/gui/executionlogwidget.cpp:
```diff
@@ -29,78 +29,73 @@
+    m_messageFilterModel->setSourceModel(m_messageModel);
+    m_messageView->setModel(m_messageFilterModel);
+    m_messageView->setContextMenuPolicy(Qt::CustomContextMenu);
+    connect(m_messageView, &LogListView::customContextMenuRequested, this, [this](const QPoint &pos) {
```
Fix coding style.

jagannatharjun replied:
I think I need to put the braces in the new line.
done.

glassez commented on src/gui/executionlogwidget.cpp:
```diff
@@ -29,78 +29,73 @@
+
+    m_peerView->setModel(m_peerModel);
+    m_peerView->setContextMenuPolicy(Qt::CustomContextMenu);
+    connect(m_peerView, &LogListView::customContextMenuRequested, this, [this](const QPoint &pos) {
```
Fix coding style.

jagannatharjun replied:
I think I need to put the braces in the new line.
done.

glassez commented on src/gui/executionlogwidget.cpp:
```diff
@@ -29,78 +29,73 @@
 }
 
-void ExecutionLogWidget::showMsgTypes(const Log::MsgTypes &types)
+void ExecutionLogWidget::setMsgTypes(const Log::MsgTypes types)
```
Please use "message" instead of "msg" whenever possible.

jagannatharjun replied:
changed the member function of `ExecutionLogWidget`, do I need to change the member function of `MainWindow` as well (`MainWindow::setExecutionLogMsgTypes` and `MainWindow::executionLogMsgTypes` )

glassez replied:
I don't insist that you change anything except the ones you touch anyway.

glassez commented on src/gui/executionlogwidget.h:
```diff
@@ -30,33 +30,39 @@
+    LogMessageModel *m_messageModel;
+    LogPeerModel *m_peerModel;
+    LogFilterModel *m_messageFilterModel;
+    LogListView *m_messageView;
```
Seems you don't need fields for views.

jagannatharjun replied:
If I understand it correctly by that logic I don't need to store LogMessageModel and LogPeerModel as well, so I removed them as well

jagannatharjun commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,124 @@
+            QStyledItemDelegate::paint(painter, option, index); // paints background, focus rect and selection rect
+
+            const QStyle *style = option.widget ? option.widget->style() : QApplication::style();
+            const int textHMargin = style->pixelMetric(QStyle::PM_FocusFrameHMargin, &option, option.widget) + 1;
```
qt guys add too
https://github.com/qt/qtbase/blob/c9b2528a9d60561e3050a3dadd86b893cd35349d/src/widgets/styles/qcommonstyle.cpp#L1019

selected item's text will be too close to focus rect

Chocobo1 commented on src/gui/executionlogwidget.cpp:
```diff
@@ -29,78 +29,75 @@
 
 #include "base/global.h"
-#include "loglistwidget.h"
+#include "base/logger.h"
```
you already included this header in executionlogwidget.h so you don't need it here.

jagannatharjun replied:
removed seems like `base/global.h` was not needed too, removed as well.

Chocobo1 commented on src/gui/executionlogwidget.h:
```diff
@@ -30,33 +30,33 @@
 public:
-    ExecutionLogWidget(QWidget *parent, const Log::MsgTypes &types);
-    void showMsgTypes(const Log::MsgTypes &types);
+    ExecutionLogWidget(QWidget *parent, Log::MsgTypes types);
```
It would be nice to also switch the parameters order: `ExecutionLogWidget(Log::MsgTypes types, QWidget *parent);`

jagannatharjun replied:
done.

Chocobo1 commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,124 @@
+
+namespace
+{
+    static const QString SEPERATOR = QStringLiteral(" - ");
```
`static` is now redundant.

jagannatharjun replied:
removed

glassez replied:
Please don't answer on each comment unless you really have to say something more meaningful than "done", "removed", "fixed" etc... You're clogging up this thread and our mailboxes.

jagannatharjun replied:
sorry, will not do anymore. 😊

glassez replied:
Also, please do not push the slightest changes to the github. When you do this every few minutes, TravisCI goes crazy! This also clogs our mailboxes with useless notifications.

Chocobo1 commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,124 @@
+            style->drawItemText(painter, textRect.adjusted(timeWidth, 0, 0, 0), 0, option.palette, (option.state & QStyle::State_Enabled), SEPERATOR);
+
+            const QVariant foreground = index.data(Qt::ForegroundRole);
+            if (foreground.canConvert<QColor>()) {             // peermodel may give invalid foreground role
```
nitpick, please trim the whitespaces before the comment mark a bit..

jagannatharjun replied:
done?

Chocobo1 commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,124 @@
+    public:
+        using QStyledItemDelegate::QStyledItemDelegate;
+
+        void paint(QPainter *painter, const QStyleOptionViewItem &option, const QModelIndex &index) const override
```
this function can be `private` isn't?

jagannatharjun replied:
made private, turns out both paint and sizehint can be private.

Chocobo1 commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,124 @@
+
+        QSize sizeHint(const QStyleOptionViewItem &option, const QModelIndex &index) const override
+        {
+            return QSize(option.fontMetrics.width(QString::fromLatin1("%1%2%3")
```
I would extract the QString into a variable to avoid this line becoming too long.

jagannatharjun replied:
done,

Chocobo1 commented on src/gui/log/logmodel.h:
```diff
@@ -0,0 +1,97 @@
+    void addNewMessage(Message message);
+
+private:
+    std::deque<Message> m_messages; // when logger is full, pop_front may take considerable time on vector like container
```
This is not good enough, you should use circular buffer (same one from logger) as suggested before.

jagannatharjun replied:
boost::circular_buffer_optimesed causing crashing 
![image](https://user-images.githubusercontent.com/34717789/78901916-a62ef600-7a96-11ea-9bd2-b402b6252140.png)

can't seem to understand why, this happens at end i.e closing the main window, std::deque works fine.

glassez replied:
>can't seem to understand why

You should at least trace it to realize what code point couses the crash.

jagannatharjun replied:
I did trace it, it was here 
https://github.com/qbittorrent/qBittorrent/blob/43e5e242ff31da009c67e4365925e11604131980/src/gui/mainwindow.cpp#L792

I spent an hour on this, only after that I posted here.

glassez replied:
I suppose it's due to incorrect order of class destructions. First views should be destroyed and then models they use.

glassez replied:
Aren't you able to achieve more depth? Try setting breakpoints in your class destructors and see what happens, in what order they are removed, and how they interact with each other.

jagannatharjun replied:
hmm changing peer model and peer view initialization order seems to fixed it thanks.

but  I have another problem 
```
void BaseLogModel::addNewMessage(const BaseLogModel::Message& message)
{
    beginInsertRows(QModelIndex(), 0, 0);
    m_messages.push_back(message);
    endInsertRows();

    const int count = rowCount();
    if (count > MAX_LOG_MESSAGES) {
        const int lastMessage = count - 1;
        beginRemoveRows(QModelIndex(), lastMessage, lastMessage);
        endRemoveRows();
    }
}
```
after switching to circular buffer,
`if (count > MAX_LOG_MESSAGES) {...}` is not needed anymore right, since pushing on the filled circular buffer, will not change the size anymore and I also then don't need to `beginRemoveRows`, I can't fit this in my mind.

jagannatharjun replied:
I've pushed changes with buffer implementation, I tested with setting MAX_LOG_MESSAGES=20 and other small values.

jagannatharjun replied:
Actually the real reason of crash was that I left some std::move calls, sorry foe troubling you guys.

Chocobo1 commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,150 @@
+    if (!index.isValid())
+        return {};
+
+    const size_t msgIndex = m_messages.size() - index.row() - 1;
```
use signed integer here.

Chocobo1 commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    message.userRole = log.type;
+    addNewMessage(std::move(message));
```
I think you should omit using move semantics to keep the code base consistent, `QVariant` uses Qt implicit sharing anyway.

Chocobo1 commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,150 @@
+    }
+}
+
+void BaseLogModel::addNewMessage(BaseLogModel::Message message)
```
And the parameter will become a const reference.

Chocobo1 commented on src/gui/log/logmodel.h:
```diff
@@ -0,0 +1,97 @@
+        QVariant time;
+        QVariant message;
+        QVariant foregroundRole;
+        QVariant userRole;
```
You should use the proper type for each field instead.

jagannatharjun replied:
doesn't that'll cause too much conversion in `BaseLogModel::data` function

Chocobo1 replied:
Maybe we can stick to Qt data types that are easily convertible to QVariant?
```c++
QString time;
QString message;
QColor foregroundRole;
MsgType userRole;  // same as `Msg::MsgType`, seems unavoidable 
```

jagannatharjun replied:
I think QVariant does dynamic memory allocation and make copies, breaking implicit sharing for original data, I call it many times during painting, but I don't have any problem in changing if you insist and for sizeHint as well.

glassez replied:
I don't see any strong reason to store it differently. QVariant is good choice for its job.

Chocobo1 replied:
I don't insist to store in non-QVariant types in this struct. However I think it is important to specify what types do we intend to store in QVariant. I think maybe add some comment in code about the types or even use `Q_Assert(QVariant::type()  == some_type)` to check the correctness.

glassez replied:
I would make it to be "read only" class that accept specific types in its constructor and store them as QVariant (getters are also of type QVariant).

jagannatharjun replied:
Where should I have to add assert? One thing is that delegate assume the type of role provided by baselogmodel so maybe baselogmodel should be responsible as well. Thus we need asserts in 'BaseLogModel::addNewMessage'

glassez replied:
Use my proposal is more reliable, IMO. You'll get something like "aggregate variant". It will perform type checking at compile time.

glassez replied:
>I would make it to be "read only" class 

It means that fields should be private and have public getters to be sure they're still assigned using constructor.

glassez commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,124 @@
+
+namespace
+{
+    const QString SEPERATOR = QStringLiteral(" - ");
```
SEP**A**RATOR

glassez commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,124 @@
+            const QRect textRect = style->subElementRect(QStyle::SE_ItemViewItemText, &option, option.widget)
+                                   .adjusted(textHMargin, textVMargin, -textHMargin, -textVMargin);
+
+            QPen coloredPen = originalPen;
```
Shouldn't you save Pen just before this line?

jagannatharjun replied:
I was told to do this way https://github.com/qbittorrent/qBittorrent/pull/12144#discussion_r404950805

glassez replied:
Unfortunately the link above doesn't work for me...

jagannatharjun replied:
it works, anyway
![image](https://user-images.githubusercontent.com/34717789/78932396-06d52780-7ac5-11ea-92c0-f3b5ed424a31.png)

Chocobo1 replied:
Maybe change `const QPen originalPen` to `QPen coloredPen` and you can reuse it.

glassez replied:
If you don't need to store original/default pen then maybe you don't need any additional variable?

jagannatharjun replied:
I need originalPen to paint separator

glassez replied:
Well, so why not save original pen immediately before you change it?

glassez commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,124 @@
+            style->drawItemText(painter, textRect.adjusted(timeWidth, 0, 0, 0), 0, option.palette, (option.state & QStyle::State_Enabled), SEPERATOR);
+
+            const QVariant foreground = index.data(Qt::ForegroundRole);
+            if (foreground.canConvert<QColor>()) { // peermodel may give invalid foreground role
```
>// peermodel may give invalid foreground role

Is it intentional behavior? What reason it does for?

jagannatharjun replied:
for the peer model we paint with default font color, I already changed the pen to `originalPen` for the painting of Separator so when this condition fails the following message will be painted with the default font color.

I think I should change it to 
// peermodel **will** give invalid foreground role

jagannatharjun replied:
ok it seems originally we paint peer messages with red color will change it accordingly

glassez commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,124 @@
+    QStringList list;
+    const QModelIndexList selectedIndexes = selectionModel()->selectedRows();
+    for (const QModelIndex &index : selectedIndexes) {
+        list.append(QString::fromLatin1("%1%2%3").arg(index.data(BaseLogModel::TimeRole).toString(), SEPERATOR
```
I would have separator inside format string to avoid yet another argument insertion. @Chocobo1?

glassez commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,157 @@
+
+BaseLogModel::BaseLogModel(QObject *object)
+    : QAbstractListModel(object)
+    , m_messages(MAX_LOG_MESSAGES)
```
How about having less message count to be displayed and not follow Logger limit?
@Chocobo1?

Chocobo1 replied:
GUI widgets should have its own display limit.
Also I would prefer larger (around 50000) if it doesn't drag the performance.

glassez replied:
>Also I would prefer larger (around 50000) if it doesn't drag the performance.

Maybe it would be better to make it configurable?
Anyway the good first step is just declare another const for GUI stuff (even if it has the same value initially).

Chocobo1 replied:
>Maybe it would be better to make it configurable?

Maybe in a later PR to avoid delaying this one...

>Anyway the good first step is just declare another const for GUI stuff (even if it has the same value initially).

Agreed.

jagannatharjun replied:
What should be the const name? Can't think of one. MAX_GUI_LOG_MESSAGES?

glassez replied:
MAX_VISIBLE_MESSAGES?
Also seems Logger constant should be private.

glassez commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,157 @@
+    if (!index.isValid())
+        return {};
+
+    const int msgIndex = m_messages.size() - index.row() - 1;
```
messageIndex

glassez commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,157 @@
+        return {};
+
+    const int msgIndex = m_messages.size() - index.row() - 1;
+    if ((msgIndex < 0) || (msgIndex >= m_messages.size()))
```
Index cannot be greater or equal to messages count due to expression above, isn't it?

glassez commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,157 @@
+{
+    for (const Log::Msg &msg : asConst(Logger::instance()->getMessages()))
+        handleNewMessage(msg);
+    connect(Logger::instance(), &Logger::newLogMessage, this, &LogMessageModel::handleNewMessage, Qt::QueuedConnection);
```
Why do you use QueuedConnection?

jagannatharjun replied:
the `newLogMessage` is emitted from `addMessage` which can be called from different threads, so I guess this way we don't need any synchronization.

Chocobo1 replied:
It would be nice to add some comment in code.

glassez replied:
Really signal receiver should know nothing about emitter internals!
You should use `Qt::QueuedConnection` directly only in case you want `QueuedConnection` explicitly for some reason. In this case you can safetely use default `Qt::AutoConnection`. Doc says about it:
>If the receiver lives in the thread that emits the signal, `Qt::DirectConnection` is used. Otherwise, `Qt::QueuedConnection` is used. The connection type is determined when the signal is emitted.

Anyway you should remove the comment above because it creates the false impression that we have an incorrect application design.

jagannatharjun replied:
AFAIK this automatic detection will only work when we use `QThread`

glassez commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,157 @@
+{
+    const QString time = QDateTime::fromMSecsSinceEpoch(peer.timestamp).toString(Qt::SystemLocaleShortDate);
+    const QString message = (peer.blocked
+                             ? tr("%1 was blocked %2", "x.y.z.w was blocked").arg(peer.ip, peer.reason)
```
Format string looks ugly.

jagannatharjun replied:
I changed it to what it was originally in `ExecutionLogWidget`, but don't know what to do further.

glassez commented on src/gui/log/logmodel.h:
```diff
@@ -0,0 +1,97 @@
+        MessageRole
+    };
+    
+    BaseLogModel(QObject *object);
```
explicit

And I would preserve default value for parameter as QObject has it.

glassez commented on src/gui/log/logmodel.h:
```diff
@@ -0,0 +1,97 @@
+    QVariant data(const QModelIndex &index, int role = Qt::DisplayRole) const override;
+
+public slots:
+    void reset();
```
Do you really need it to be declared as slot? If no, you even can omit Q_OBJECT.

glassez commented on src/gui/log/logmodel.h:
```diff
@@ -0,0 +1,97 @@
+        QVariant userRole;
+    };
+
+    void addNewMessage(const Message& message);
```
Broken coding style (ampersand position).

glassez commented on src/gui/log/logmodel.h:
```diff
@@ -0,0 +1,97 @@
+
+class LogMessageModel : public BaseLogModel
+{
+    Q_OBJECT
```
Don't forget to add Q_DISABLE_COPY to all QObject subclasses.

glassez commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,124 @@
+
+        QSize sizeHint(const QStyleOptionViewItem &option, const QModelIndex &index) const override
+        {
+            const QString displayText = QString::fromLatin1("%1%2%3").arg(index.data(BaseLogModel::TimeRole).toString(), SEPERATOR
```
Seems you can extract "build display text" logic into function to avoid code duplication.

jagannatharjun commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,123 @@
+
+            painter->setPen(originalPen);
+            const QFontMetrics fontMetrics = painter->fontMetrics(); // option.fontMetrics adds extra padding to QFontMetrics::width
+            const int timeWidth = fontMetrics.width(time);
```
![image](https://user-images.githubusercontent.com/34717789/78940860-2a539e80-7ad4-11ea-9e19-03c1d573fa5e.png)

Chocobo1 commented on src/base/logger.cpp:
```diff
@@ -72,7 +72,7 @@ void Logger::freeInstance()
 {
     QWriteLocker locker(&m_lock);
-    const Log::Msg msg = {m_msgCounter++, QDateTime::currentMSecsSinceEpoch(), type, message.toHtmlEscaped()};
+    const Log::Msg msg = {m_msgCounter++, QDateTime::currentMSecsSinceEpoch(), type, message};
```
why this again?

jagannatharjun replied:
I think this was done to support html formatting else where ( n LogListWidget), this is not needed anymore, if not removed here, had to do unescape in model or delegate, what do we prefer?

Chocobo1 replied:
It was made to avoid some attacks. IIRC calling `toHtmlEscaped()` here would mess up the saved log.txt right? If so, it is not preferable, we should not do that and escape it at model.

glassez replied:
Sorry, I can't clearly understand what you're talking about...
IMO, it's incorrect to escape here.

Chocobo1 replied:
disregard my comment.

jagannatharjun replied:
Should I remove these changes and unescape at Model or leave the changes?

Chocobo1 replied:
keep the the current changes. no need to make further changes.

glassez replied:
You should keep your changes here.

Chocobo1 commented on src/gui/log/logmodel.h:
```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include <QAbstractListModel>
+#include <boost/circular_buffer.hpp>
```
wrong order.

Chocobo1 commented on src/gui/log/logmodel.h:
```diff
@@ -0,0 +1,96 @@
+
+    explicit BaseLogModel(QObject *parent = nullptr);
+
+    int rowCount(const QModelIndex &parent = {}) const override;
```
some of those can be `private` right?

glassez replied:
Isn't it supposed to be accessed from outside of the class?

Chocobo1 commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,158 @@
+    }
+}
+
+void BaseLogModel::addNewMessage(const BaseLogModel::Message& message)
```
coding style.

Chocobo1 commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,123 @@
+            const QPen originalPen = painter->pen();
+            painter->save();
+
+            QStyledItemDelegate::paint(painter, option, index); // paints background, focus rect and selection rect
```
Now I think `QStyledItemDelegate::paint` should preserve the `painter` state (adhere to their own spec), so maybe we can swap it:
```c++
{ // function start
QStyledItemDelegate::paint(painter, option, index); // paints background, focus rect and selection rect

const QPen originalPen = painter->pen();
painter->save();
// ...
```

jagannatharjun replied:
We need originalPen to paint separator

Chocobo1 commented on src/gui/executionlogwidget.cpp:
```diff
@@ -29,78 +29,73 @@
 }
 
-void ExecutionLogWidget::addLogMessage(const Log::Msg &msg)
+void ExecutionLogWidget::displayContextMenu(const QPoint &pos, const LogListView *view, const BaseLogModel *model) const
```
Seems this doesn't need to be in class, it should be moved to anonymous namespace.

jagannatharjun replied:
This internally calls tr, and I was once asked to only call tr of concerned class not QObject::tr directly, otherwise my first thought was to make it free function as well.

glassez replied:
QObject::tr in free functuon is bad since it doesn't produce any meaningful context.

Chocobo1 replied:
ok, forget it.

Chocobo1 commented on src/gui/executionlogwidget.cpp:
```diff
@@ -29,78 +29,73 @@
+    LogListView *messageView = new LogListView(this);
+    messageView->setModel(m_messageFilterModel);
+    messageView->setContextMenuPolicy(Qt::CustomContextMenu);
+    connect(messageView, &LogListView::customContextMenuRequested, this, [this, messageView, messageModel](const QPoint &pos)
```
seems you don't really need to capture `this` into lambda.

Chocobo1 commented on src/gui/executionlogwidget.cpp:
```diff
@@ -29,78 +29,73 @@
+    LogListView *peerView = new LogListView(this);
+    peerView->setModel(peerModel);
+    peerView->setContextMenuPolicy(Qt::CustomContextMenu);
+    connect(peerView, &LogListView::customContextMenuRequested, this, [this, peerView, peerModel](const QPoint &pos)
```
same here.

Chocobo1 commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,158 @@
+    if (m_messages.size() == MAX_LOG_MESSAGES) {
+        const int lastMessage = m_messages.size() - 1;
+        beginRemoveRows(QModelIndex(), lastMessage, lastMessage);
+        m_messages.pop_front();
```
I *suspect* this isn't needed...

glassez replied:
At least you can try to test it without this line.

glassez replied:
But theoretically it should be here. Otherwise rowCount() still returns incorrect value after you notify about removed row.

Chocobo1 replied:
I just noticed the comment above, never mind here.

glassez commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    beginInsertRows(QModelIndex(), 0, 0);
+    m_messages.push_back(message);
```
Why not store messages in the same order as they're displayed?

glassez commented on src/gui/log/logmodel.h:
```diff
@@ -0,0 +1,96 @@
+
+private:
+    boost::circular_buffer_space_optimized<Message> m_messages;
+    static constexpr int MAX_VISIBLE_MESSAGES = MAX_LOG_MESSAGES;
```
Why not move it to anonymous namespace? "static private const" seems to be useless unless it's initialized in class constructor.

Chocobo1 commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,193 @@
+
+namespace
+{
+    constexpr int MAX_VISIBLE_MESSAGES = MAX_LOG_MESSAGES;
```
I would use `constexpr` when you really need to eval at compile time, here `const` would be enough.
And it should not be related to `MAX_LOG_MESSAGES`.

Chocobo1 commented on src/gui/log/logmodel.h:
```diff
@@ -0,0 +1,106 @@
+    class Message
+    {
+    public:
+        Message(QString time, QString message, QColor foreground, QVariant data = {});
```
it should not have any default values.

Chocobo1 replied:
All of the parameters should take const references except the last one, last one should be the original type instead of QVariant.

Chocobo1 commented on src/gui/log/logmodel.h:
```diff
@@ -0,0 +1,106 @@
+        QVariant userRole() const;
+
+    private:
+        QVariant m_time;
```
seems every one of them can be `const`

jagannatharjun replied:
I tried, circular buffer needs operator=

Chocobo1 replied:
thanks, never mind then.

Chocobo1 commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,193 @@
+        return {};
+
+    const int messageIndex = index.row();
+    if ((messageIndex > 0) && (messageIndex >= m_messages.size()))
```
suspicious...

glassez commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,192 @@
+    switch (role) {
+    case Qt::ForegroundRole:
+        return message.foregroundRole();
+    case Qt::UserRole:
```
Why didn't you make a specific name for this one (and all its stuff)? Why not "MessageTypeRole"?

glassez replied:
Also the class members have inconsistent names. Some of them have "role" suffix and the rest has no one.

jagannatharjun replied:
I removed suffixes from Message class, this should make the class independent from "role" handling,
I replaced "DataRole" with "MessageTypeRole", which includes different values for type and foreground in addition to message and time. I included the foreground role in the enum as well since QStyledItemDelegate::paint shouldn't paint anything related to the foreground. TypeRole enum constant provides better information about the type of data as compared to the previously used value of Qt::UserRole, one side effect of that was that had to include `logModel.h` in `logFilterModel.cpp`, but that shouldn't be a problem since there is a clear dependency between these two classes.

glassez replied:
Oops... unfortunately, I didn't pay enough attention to this earlier. As a result, we got an incorrect dependency here.
The model should not depend on any particular view/delegate. It must always return view-independent data at least for mandatory QDisplayRole (so that any view can display its data at least in simplified text format). You can, of course, object that we will never use this model with any other views, but this is an incorrect reason.
There is also no point in using a custom foreground role. "I included the foreground role in the enum as well since QStyledItemDelegate::paint shouldn't paint anything related to the foreground"  is an incorrect judgment. If you don't return Qt::ForegroundRole it just uses default value. If you want to prevent QStyledItemDelegate::paint from painting text you should clear text in passed parameters.

glassez replied:
>You can, of course, object that we will never use this model with any other views

By the way, we could give the user the option to select a simplified log view (just not using a custom delegate). This might improve performance a bit, right?

jagannatharjun replied:
> If you want to prevent QStyledItemDelegate::paint from painting text you should clear text in passed parameters.

That's the reason I'm not using DisplayRole, even returning for Qt::ForegroundRole will not do anything since DisplayRole is null and the foreground role we return does only make sense with the our message.
I can't return valid value for DisplayRole, since then QStyledItemDelegate will paint it with Foreground, I want `QStyledItemDelegate` to only paint background since painting those myself was a headache anyway. 
If we were to return a full string with DisplayRole(time + message), it will be useless effort to seperate them. 

> The model should not depend on any particular view/delegate.

I don't understand this logic. Probably I'm inexperience, but you yourself said, that "model should prepare data and delegate should do the painting accordingly"
the whole purpose of data is too model is to store and provide data in "convenience form", 
I think you're trying to remove the dependency between model and deleagate, but I don't think it is possible, they are made and build for eachother(XD) and with that logic you will never able to use QML.

jagannatharjun replied:
>By the way, we could give the user the option to select a simplified log view (just not using a custom delegate). This might improve performance a bit, right?


I don't see any performance improvement with that, the only extra work we do is to paint part of a string separately, that doesn't involve any performance heavy calculation or something on that line.

glassez replied:
>the whole purpose of the model is to store and provide data in "convenience form",
I think you're trying to remove the dependency between model and delegate

Well, it seems we have a misunderstanding of how Qt Model/View is supposed to work (ideally). It's time to make it clear (for me as well). I will try to do this when I find myself near something more convenient to comment on than my smartphone.

jagannatharjun replied:
IMHO, you want to have model return valid values for ForegroundRole and DisplayRole, but this will just create unnecessary work, first, we will need to separate message and time. Second, I want to use `QStyleItemDelegate::paint` but can't find a way to modify or create a new QModelIndex that fit our needs that is one without DisplayRole and ForegroundRole, 
the second one is the biggest blocker and if we had to not use QStyleItemDelegate::paint, that'll just create unnecessary code which may or may not work everywhere. 
For the record, I tried `QStyle::CE_ItemViewItem` it doesn't paint selection rectangle for items under the mouse.  
@Chocobo1 opinions?

Man, I was thinking this PR will complete today😢

glassez replied:
Guys, relax, so as not to imitate you, I will not insist on further improving the Model/View architecture in this PR. In reality, we don't have a completely perfect implementation anywhere in the code. And it seems that this is not even always possible.

glassez commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,192 @@
+
+void BaseLogModel::addNewMessage(const BaseLogModel::Message &message)
+{
+    // if row is inserted on filled up buffer, the size will not change but because of calling of beginInsertRows function
```
Please reformat this comment. The first line is too long while the secnd one is short.

Chocobo1 commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,192 @@
+#include <QColor>
+#include <QPalette>
+
+#include "base/global.h"
```
Seems it isn't in use.

jagannatharjun replied:
`asConst` is used in constructors of LogMessageModel of LogPeerModel.

glassez commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,192 @@
+    connect(Logger::instance(), &Logger::newLogMessage, this, &LogMessageModel::handleNewMessage, Qt::QueuedConnection);
+}
+
+void LogMessageModel::handleNewMessage(const Log::Msg &log)
```
"log" is incorrect name for this parameter. I would call it "message" and rename the next one to "messageText".

glassez commented on src/gui/log/logmodel.cpp:
```diff
@@ -0,0 +1,192 @@
+{
+    const QString time = QDateTime::fromMSecsSinceEpoch(peer.timestamp).toString(Qt::SystemLocaleShortDate);
+    const QString message = peer.blocked
+            ? tr("%1 was blocked %2", "0.0.0.0 was blocked due to reason").arg(peer.ip, peer.reason)
```
Format should be either "%1 was blocked due to %2" or "%1 was blocked. Reason: %2".

glassez commented:
>AFAIK this automatic detection will only work when we use `QThread`

What's the problem? What exactly is bothering you?

jagannatharjun commented:
> What's the problem? What exactly is bothering you?

the newLogMessage is emitted from addMessage which can be called from different threads

glassez commented:
>the newLogMessage is emitted from addMessage which can be called from different threads

Yes. There should be no problem. According to the docs (I posted in my comment above):
>If the receiver lives in the thread that emits the signal, `Qt::DirectConnection` is used. Otherwise, `Qt::QueuedConnection` is used. **The connection type is determined when the signal is emitted**.

jagannatharjun commented:
Ok now I get it.

Chocobo1 commented:
@jagannatharjun 
Please fix all `extra ‘;’` warnings: https://travis-ci.org/github/qbittorrent/qBittorrent/jobs/673732414

Chocobo1 commented:
And those:
```
../../src/gui/log/loglistview.cpp: In member function ‘virtual void {anonymous}::LogItemDelegate::paint(QPainter*, const QStyleOptionViewItem&, const QModelIndex&) const’:
../../src/gui/log/loglistview.cpp:79:57: warning: ‘int QFontMetrics::width(const QString&, int) const’ is deprecated: Use QFontMetrics::horizontalAdvance [-Wdeprecated-declarations]
   79 |             const int timeWidth = fontMetrics.width(time);
      |                                                         ^
In file included from /usr/include/qt/QtWidgets/qwidget.h:50,
                 from /usr/include/qt/QtWidgets/qframe.h:44,
                 from /usr/include/qt/QtWidgets/qabstractscrollarea.h:44,
                 from /usr/include/qt/QtWidgets/qabstractitemview.h:44,
                 from /usr/include/qt/QtWidgets/qlistview.h:44,
                 from /usr/include/qt/QtWidgets/QListView:1,
                 from ../../src/gui/log/loglistview.h:32,
                 from ../../src/gui/log/loglistview.cpp:30:
/usr/include/qt/QtGui/qfontmetrics.h:106:9: note: declared here
  106 |     int width(const QString &, int len = -1) const;
      |         ^~~~~
../../src/gui/log/loglistview.cpp:84:100: warning: ‘int QFontMetrics::width(const QString&, int) const’ is deprecated: Use QFontMetrics::horizontalAdvance [-Wdeprecated-declarations]
   84 |             style->drawItemText(painter, textRect.adjusted((timeWidth + fontMetrics.width(SEPARATOR)), 0, 0, 0), 0
      |                                                                                                    ^
In file included from /usr/include/qt/QtWidgets/qwidget.h:50,
                 from /usr/include/qt/QtWidgets/qframe.h:44,
                 from /usr/include/qt/QtWidgets/qabstractscrollarea.h:44,
                 from /usr/include/qt/QtWidgets/qabstractitemview.h:44,
                 from /usr/include/qt/QtWidgets/qlistview.h:44,
                 from /usr/include/qt/QtWidgets/QListView:1,
                 from ../../src/gui/log/loglistview.h:32,
                 from ../../src/gui/log/loglistview.cpp:30:
/usr/include/qt/QtGui/qfontmetrics.h:106:9: note: declared here
  106 |     int width(const QString &, int len = -1) const;
      |         ^~~~~
../../src/gui/log/loglistview.cpp: In member function ‘virtual QSize {anonymous}::LogItemDelegate::sizeHint(const QStyleOptionViewItem&, const QModelIndex&) const’:
../../src/gui/log/loglistview.cpp:91:65: warning: ‘int QFontMetrics::width(const QString&, int) const’ is deprecated: Use QFontMetrics::horizontalAdvance [-Wdeprecated-declarations]
   91 |             return QSize(option.fontMetrics.width(logText(index)), QStyledItemDelegate::sizeHint(option, index).height());
      |                                                                 ^
In file included from /usr/include/qt/QtWidgets/qwidget.h:50,
                 from /usr/include/qt/QtWidgets/qframe.h:44,
                 from /usr/include/qt/QtWidgets/qabstractscrollarea.h:44,
                 from /usr/include/qt/QtWidgets/qabstractitemview.h:44,
                 from /usr/include/qt/QtWidgets/qlistview.h:44,
                 from /usr/include/qt/QtWidgets/QListView:1,
                 from ../../src/gui/log/loglistview.h:32,
                 from ../../src/gui/log/loglistview.cpp:30:
/usr/include/qt/QtGui/qfontmetrics.h:106:9: note: declared here
  106 |     int width(const QString &, int len = -1) const;
      |         ^~~~~
```

```
../../src/gui/log/logmodel.cpp: In member function ‘virtual QVariant BaseLogModel::data(const QModelIndex&, int) const’:
../../src/gui/log/logmodel.cpp:94:45: warning: comparison of integer expressions of different signedness: ‘const int’ and ‘boost::circular_buffer<BaseLogModel::Message, std::allocator<BaseLogModel::Message> >::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
   94 |     if ((messageIndex < 0) || (messageIndex >= m_messages.size()))
      |                                ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
```
I would use a static_cast<int> on m_messages.size().

Chocobo1 commented:
The "g" character bottom is cut off: ![screenshot](https://user-images.githubusercontent.com/9395168/79092215-c0145580-7d82-11ea-9fc6-201a4b602f8e.png)

Chocobo1 commented:
Current master:
![before](https://user-images.githubusercontent.com/9395168/79092677-35345a80-7d84-11ea-9a93-1436fc04ccfe.png)
This PR:
![after](https://user-images.githubusercontent.com/9395168/79092680-36658780-7d84-11ea-8b10-11af65cb75a1.png)
Notice in this PR, the selection bar has different height margins for top and bottom.

Chocobo1 commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,122 @@
+
+            const QPen originalPen = painter->pen();
+            QPen coloredPen = originalPen;
+            coloredPen.setColor(Qt::gray);
```
Seems this is not the same gray, could you preserve the old value?

Chocobo1 replied:
Please try `Qt::darkGray`.

jagannatharjun commented:
>    79 |             const int timeWidth = fontMetrics.width(time);

`QFontMetrics::horizontalAdvance` is not available till 5.11 had to use ifdef.

> The "g" character bottom is cut off:

are you sure you're using current version, I thought I fixed this 
```
            const QRect textRect = style->subElementRect(QStyle::SE_ItemViewItemText, &option, option.widget)
                                   .adjusted(textHMargin, textVMargin, 0, 0);
``
I can replicate with or without custom theme (widows), can you try with textVMargin as zero?
here https://github.com/qt/qtbase/blob/c9b2528a9d60561e3050a3dadd86b893cd35349d/src/widgets/styles/qcommonstyle.cpp#L1019
they don't adjust for vertical margin, but I was getting not centered results on my machine.

Chocobo1 commented:
>are you sure you're using current version, I thought I fixed this

I check out just before testing it, should be the latest.

>I can't replicate with or without custom theme (windows), can you try with textVMargin as zero?

Tested, now there is no cut off on character bottom. But both top/bottom margin still seems to be zero now.

>EDIT: can you try with const QStyleOptionViewItem *option as nullptr in pixelMetric calls.

I don't understand. You'll need to be more specific.

jagannatharjun commented:
replicated on ubuntu, will try to fix it.

jagannatharjun commented:
![Uploading image.png…]()

@Chocobo1 Can you try too, it should be fixed now.

Chocobo1 commented:
>Please have a look, it should be fixed now.

I confirm, it is alright now.

Chocobo1 commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,133 @@
+
+        QSize sizeHint(const QStyleOptionViewItem &option, const QModelIndex &index) const override
+        {
+            return QSize(horizontalAdvance(option.fontMetrics, logText(index))
```
omit `QSize` here: `return {...};`

Chocobo1 commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,133 @@
+        QSize sizeHint(const QStyleOptionViewItem &option, const QModelIndex &index) const override
+        {
+            return QSize(horizontalAdvance(option.fontMetrics, logText(index))
+                         , option.fontMetrics.height() + 2 * QApplication::style()->pixelMetric(QStyle::PM_FocusFrameVMargin, &option, option.widget));
```
Too long.. use a variable for it.

jagannatharjun commented:
on Windows.
current master
![image](https://user-images.githubusercontent.com/34717789/79099128-14b6d100-7d81-11ea-94aa-50ece28494a4.png)

this PR
![qbit-log](https://user-images.githubusercontent.com/34717789/79099117-0e285980-7d81-11ea-8ca1-b5e8640e9823.JPG)

acceptable?

jagannatharjun commented:
> there is extra space at the bottom

seems like adjusted call on textRect was unnecessary

jagannatharjun commented:
![image](https://user-images.githubusercontent.com/34717789/79111672-99621900-7d9a-11ea-9075-988011a4267f.png)
![image](https://user-images.githubusercontent.com/34717789/79111727-b696e780-7d9a-11ea-913b-c42b01715717.png)
![image](https://user-images.githubusercontent.com/34717789/79111829-e7771c80-7d9a-11ea-9748-5d577d4d9efb.png)

jagannatharjun commented:
fonts are not scaling with windows scale setting 

![image](https://user-images.githubusercontent.com/34717789/79125209-65e3b680-7dbb-11ea-834d-51e7a1d7819a.png)

hmm

glassez commented:
>fonts are not scaling with windows scale setting

Has QStyledItemDelegate the same problem?

jagannatharjun commented:
> Has QStyledItemDelegate the same problem?

that's the weird part, following code will just work, even don't require setting text margin that I do in `LogItemDelegate::paint`, everything will just work in all platforms
even if I put the whole `LogItemDelegate::paint` in the following code, it will just work, that's what giving me a headache.

```c++
#include <QAbstractListModel>
#include <QApplication>
#include <QDebug>
#include <QListView>
#include <QPainter>
#include <QStyle>
#include <QStyledItemDelegate>

class Delegate : public QStyledItemDelegate {
 public:
  using QStyledItemDelegate::QStyledItemDelegate;

  void paint(QPainter *painter, const QStyleOptionViewItem &option, const QModelIndex &index) const override {
    painter->save();
    QStyledItemDelegate::paint(painter, option, index);
    const QStyle *style = option.widget ? option.widget->style() : QApplication::style();
    const QRect textRect = style->subElementRect(QStyle::SE_ItemViewItemText, &option, option.widget);

    style->drawItemText(painter, textRect, 0, option.palette, (option.state & QStyle::State_Enabled),
                        index.data(Qt::UserRole + 1).toString());
    painter->restore();
  }
};

class Model : public QAbstractListModel {
  QVector<QString> strs;

 public:
  int rowCount(const QModelIndex &parent = {}) const override { return strs.size(); }
  int columnCount(const QModelIndex &parent = {}) const override { return 1; }
  QVariant data(const QModelIndex &index, int role = Qt::DisplayRole) const override {
    if (index.isValid() && role == Qt::UserRole + 1) return strs[index.row()];
    return {};
  }
  void add(QString s) {
    beginInsertRows(QModelIndex(), 0, 0);
    strs.append(s);
    endInsertRows();
  }
};

int main(int argc, char *argv[]) {
  QApplication a(argc, argv);

  QListView view;
  Model model;
  view.setItemDelegate(new Delegate(&view));
  view.setModel(&model);
  model.add("Text 1");
  model.add("Text 1");
  model.add("Text 1");
  model.add("Text 1");

  view.show();
  return a.exec();
}

```

Chocobo1 commented:
>fonts are not scaling with windows scale setting

Which Qt version were you using?

jagannatharjun commented:
> Which Qt version were you using?

using 5.14.1

with 5.12.5 ->
make text bigger -> 225%
![image](https://user-images.githubusercontent.com/34717789/79207784-c677ff80-7e5e-11ea-83d6-7dea43ef0bfb.png)

scaling 175%
![image](https://user-images.githubusercontent.com/34717789/79207950-050dba00-7e5f-11ea-973a-520e530a55e9.png)

TransferListWidget scale well in both scenarios and both qt versions

jagannatharjun commented:
scalling 175%
![headache](https://user-images.githubusercontent.com/34717789/79245450-bcbdbe80-7e95-11ea-85fe-b70adc1e5195.JPG)
![sol](https://user-images.githubusercontent.com/34717789/79245463-c2b39f80-7e95-11ea-8263-784b2c931c05.JPG)

"make text size bigger" 225%
![textscaling](https://user-images.githubusercontent.com/34717789/79245516-d2cb7f00-7e95-11ea-93cb-c5b2c5a8cb44.JPG)

@Chocobo1 @glassez please have a look

jagannatharjun commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,140 @@
+
+            QFont font = option.font;
+            if (option.font.pointSize() > 0)
+                font.setPointSize(option.font.pointSize()); // somehow this needs to be set directly else painter will use default font
```
https://github.com/qbittorrent/qBittorrent/pull/12458#issuecomment-613523552

Chocobo1 replied:
This sounds better to me: "... else painter ..." -> "... otherwise painter ..."

Chocobo1 commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,140 @@
+
+        QSize sizeHint(const QStyleOptionViewItem &option, const QModelIndex &index) const override
+        {
+            const QSize fontSize = option.fontMetrics.size(0, logText(index)) + QSize(4, 4);
```
`QSize(4, 4)` where does it come from? Please add comment to code.

Chocobo1 commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,140 @@
+        {
+            const QSize fontSize = option.fontMetrics.size(0, logText(index)) + QSize(4, 4);
+            const QSize defaultSize = QStyledItemDelegate::sizeHint(option, index);
+            const QSize margins = QSize(std::max(0, defaultSize.width() - fontSize.width())
```
`const QSize margins = (defaultSize - fontSize).expandedTo({0, 0});`

Chocobo1 commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,140 @@
+            QFont font = option.font;
+            if (option.font.pointSize() > 0)
+                font.setPointSize(option.font.pointSize()); // somehow this needs to be set directly else painter will use default font
+            painter->setFont(font);
```
This should be equivalent `painter->setFont(option.font);`, could you please test it again?

jagannatharjun replied:
![image](https://user-images.githubusercontent.com/34717789/79306288-9722ca80-7f12-11ea-825b-070a38f6be1a.png)

Chocobo1 commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,140 @@
+
+            painter->setPen(originalPen);
+            const QFontMetrics fontMetrics = painter->fontMetrics(); // option.fontMetrics adds extra padding to QFontMetrics::width
+            const int separatorLeft = horizontalAdvance(fontMetrics, time);
```
seems the variable is badly named, I would use `separatorWidth`

jagannatharjun replied:
It is "separatorleft", and following one is "messageLeft"

Chocobo1 replied:
Hmm... I see what you mean. But "left" isn't a meaningful thing, it is actually a coordinate.

Chocobo1 replied:
maybe `separatorCoordinateX`?

Chocobo1 commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,140 @@
+
+            coloredPen.setColor(index.data(BaseLogModel::ForegroundRole).value<QColor>());
+            painter->setPen(coloredPen);
+            const int messageLeft = (separatorLeft + horizontalAdvance(fontMetrics, SEPARATOR));
```
same here.

Chocobo1 commented on src/gui/log/loglistview.cpp:
```diff
@@ -88,24 +88,24 @@ namespace
         QSize sizeHint(const QStyleOptionViewItem &option, const QModelIndex &index) const override
         {
-            const QSize fontSize = option.fontMetrics.size(0, logText(index)) + QSize(4, 4);
+            static const QSize minimumFontPadding(4, 4);
```
why the `static`? It can be removed IMO.

glassez replied:
Or declare it in anonymous namespace above...

glassez commented on src/gui/log/loglistview.cpp:
```diff
@@ -0,0 +1,140 @@
+
+            QFont font = option.font;
+            if (option.font.pointSize() > 0)
+                font.setPointSize(option.font.pointSize()); // somehow this needs to be set directly otherwise painter will use default font
```
It still seems weird... if copy constructor doesn't produce the same "value" there is bug in it.

jagannatharjun replied:
I couldn't get if I was doing something fishy, everything is const, so I don't think I can do anything here then
I tried checking qt sources for qfont but they are using private impl so I didn't try to further delve into it.

jagannatharjun replied:
I tried sanitizer, both master and this branch seems to be throwing same warnings (there were quite a number of warnings especially with thread sanitizer)

glassez replied:
I wonder how QStyledItemDelegate draws the text?

glassez replied:
But it's most likely a problem with QFont since an action that shouldn't have any effect still has it.
Just for fun... can you test the following?
```c++
font.setPointSize(font.pointSize());
```

jagannatharjun replied:
hmm it works 
![image](https://user-images.githubusercontent.com/34717789/79464484-bce5c780-8017-11ea-9733-83405299b002.png)

Q_ASSERT(font.pointSize() != option.font.pointSize());
actually fails

Chocobo1 commented:
@jagannatharjun 
Thank you!
